xrootd/encoder/encoder.go

```diff
@@ -0,0 +1,178 @@
+	pos := 0
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		fieldSize := 0
+		switch field.Kind() {
+		case reflect.Uint8:
+			fieldSize = 1
+			data[pos] = uint8(field.Uint())
+		case reflect.Uint16:
+			fieldSize = 2
+			binary.BigEndian.PutUint16(data[pos:pos+fieldSize], uint16(field.Uint()))
+		case reflect.Int32:
+			fieldSize = 4
+			binary.BigEndian.PutUint32(data[pos:pos+fieldSize], uint32(field.Int()))
+		case reflect.Int64:
+			fieldSize = 8
+			binary.BigEndian.PutUint64(data[pos:pos+fieldSize], uint64(field.Int()))
+		case reflect.Slice:
+			elementKind := field.Type().Elem().Kind()
+			if elementKind != reflect.Uint8 {
+				return nil, errors.Errorf("Cannot marshal slice of %s, only slices of uint8 are supported", elementKind)
+			}
+
+			fieldSize = field.Len()
+			reflect.Copy(reflect.ValueOf(data[pos:pos+fieldSize]), field)
+		case reflect.Array:
+			elementKind := field.Type().Elem().Kind()
+			if elementKind != reflect.Uint8 {
+				return nil, errors.Errorf("Cannot marshal array of %s, only arrays of uint8 are supported", elementKind)
+			}
+
+			fieldSize = field.Len()
+			reflect.Copy(reflect.ValueOf(data[pos:pos+fieldSize]), field)
+
+		default:
+			return nil, errors.Errorf("Cannot marshal kind %s", field.Kind())
+		}
+		pos += fieldSize
+	}
+
+	return data, nil
+}
+
+// Unmarshal decodes data from byte slice following the XRootd protocol specification.
+// Fields are decoded in the same order as they are specified in the struct definition.
+// Each field is decoded from network byte order (BigEndian) without any alignment or padding.
+// Slices and arrays of uint8 are binary copied without further decoding.
+// Since the length of the slice is unknown, all bytes to the end of data is copied to it.
+// Supported types are: uint8, uint16, int32, int64, slices and arrays of uint8.
+func Unmarshal(data []byte, x interface{}) (err error) {
+	v := reflect.ValueOf(x)
+
+	if v.Kind() == reflect.Ptr {
+		v = v.Elem()
+	}
+
+	if v.Kind() != reflect.Struct {
+		return errors.Errorf("Cannot unmarshal %s, struct is expected", v.Kind())
+	}
+
+	pos := 0
+
+	for i := 0; i < v.NumField() && err == nil; i++ {
+		field := v.Field(i)
+		fieldSize := 0
+		switch field.Kind() {
+		case reflect.Uint8:
+			fieldSize = 1
+			field.SetUint(uint64(data[pos]))
+		case reflect.Uint16:
+			fieldSize = 2
+			var value = binary.BigEndian.Uint16(data[pos : pos+fieldSize])
+			field.SetUint(uint64(value))
+		case reflect.Int32:
+			fieldSize = 4
+			var value = int32(binary.BigEndian.Uint32(data[pos : pos+fieldSize]))
+			field.SetInt(int64(value))
+		case reflect.Int64:
+			fieldSize = 8
+			var value = int64(binary.BigEndian.Uint64(data[pos : pos+fieldSize]))
+			field.SetInt(value)
+		case reflect.Slice:
+			elementKind := field.Type().Elem().Kind()
+			if elementKind != reflect.Uint8 {
+				return errors.Errorf("Cannot unmarshal slice of %s, only slices of uint8 are supported", elementKind)
+			}
+
+			bytes := data[pos:]
```

Comment:
hum... if I am not mistaken...
this assumes that a slice is always the last field of a struct.
this also assumes we'll never have 2 fields that are slices.

is this really the case?
naively, I would have assumed that the following C++ payload:
```c
struct Request {
    int32_t msgLen;
    char *msg;
};
```

would translate into:
```go
type Request struct {
    Msg []byte
}
```

with the following layout on the wire: `[4 bytes: msgLen][msgLen bytes: msg]`.

so the `Marshal` function would - when encountering a `[]byte` - first write the length of the slice on the wire, and then copy the whole slice.
similarly, the `Unmarshal` function would - when encountering a `[]byte` field - first read a `uint32` (the length) and read as many bytes.

but maybe I am missing something?

Reply:
Most of time it's so, but on some requests / responses length is not specified before bytes.
For example, read with readahead, see http://xrootd.org/doc/dev45/XRdv310.pdf (page 99) - rvec's length is passed indirectly via alen.

Also, in responses, for example read response, such scheme exists:
// Start of the header
kXR_char streamid[2]
kXR_unt16 status
kXR_int32 dlen  // length of the response
// End of the header
// Start of data that will be passed to the unmarshal
kXR_char data[dlen]
// End of data that will be passed to the unmarshal

However, I don't remember any requests/responses that have 2 slices of bytes. I'll look into specification once more and check if it's truth.

Any suggestions are welcome! :)